plutus-tx-plugin/test/size/Main.hs

```diff
@@ -0,0 +1,222 @@
+{-# LANGUAGE DataKinds       #-}
+{-# LANGUAGE KindSignatures  #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Main (main) where
+
+import Data.Kind (Type)
+import Data.Tagged (Tagged (Tagged))
+import PlutusTx.Code (CompiledCode, sizePlc)
+import PlutusTx.IsData.Class (fromBuiltinData, toBuiltinData, unsafeFromBuiltinData)
+import PlutusTx.Prelude qualified as Plutus
+import PlutusTx.Ratio qualified as PlutusRatio
+import PlutusTx.TH (compile)
+import Prelude
+import Test.Tasty (TestTree, defaultMain, testGroup)
+import Test.Tasty.Providers (IsTest (run, testOptions), singleTest, testFailed, testPassed)
+import Type.Reflection (Typeable)
+
+main :: IO ()
+main = defaultMain . testGroup "Size regression tests" $ [
+  testGroup "Rational" [
+    testGroup "Eq" [
+      fitsInto "==" ratEq 80,
+      fitsInto "/=" ratNeq 90
+      ],
+    testGroup "Ord" [
+      fitsInto "compare" ratCompare 129,
+      fitsInto "<=" ratLe 67,
+      fitsInto ">=" ratGe 67,
+      fitsInto "<" ratLt 67,
+      fitsInto ">" ratGt 67,
+      fitsInto "max" ratMax 77,
+      fitsInto "min" ratMin 77
+      ],
+    testGroup "Additive" [
+      fitsInto "+" ratPlus 186,
+      fitsInto "zero" ratZero 20,
+      fitsInto "-" ratMinus 186,
+      fitsInto "negate (specialized)" ratNegate 32
+      ],
+    testGroup "Multiplicative" [
+      fitsInto "*" ratTimes 178,
+      fitsInto "one" ratOne 23,
+      fitsInto "scale" ratScale 127
+      ],
+    testGroup "Serialization" [
+      fitsInto "toBuiltinData" ratToBuiltin 87,
+      fitsInto "fromBuiltinData" ratFromBuiltin 471,
+      fitsInto "unsafeFromBuiltinData" ratUnsafeFromBuiltin 299
+      ],
+    testGroup "Construction" [
+      fitsInto "unsafeRatio" ratMkUnsafe 185,
+      fitsInto "ratio" ratMkSafe 292,
+      fitsInto "fromInteger" ratFromInteger 21
+      ],
+    testGroup "Other" [
+      fitsInto "numerator" ratNumerator 24,
+      fitsInto "denominator" ratDenominator 24,
+      fitsInto "round" ratRound 390,
+      fitsInto "truncate" ratTruncate 28,
+      fitsInto "properFraction" ratProperFraction 61,
+      fitsInto "recip" ratRecip 101,
+      fitsInto "abs (specialized)" ratAbs 76
+      ],
+    testGroup "Comparison" [
+      fitsUnder "negate" ("specialized", ratNegate) ("general", genNegate),
+      fitsUnder "abs" ("specialized", ratAbs) ("general", genAbs),
+      fitsUnder "scale" ("type class method", ratScale) ("equivalent in other primitives", genScale)
+      ]
+    ]
+  ]
+
+-- Helpers
+
+-- Size testing for Tasty
+
+fitsInto :: forall (a :: Type) .
+  (Typeable a) =>
+  String -> CompiledCode a -> Integer -> TestTree
+fitsInto name cc = singleTest name . SizeTest cc
+
+data SizeTest (a :: Type) = SizeTest (CompiledCode a) Integer
+
+instance (Typeable a) => IsTest (SizeTest a) where
```

Comment:
Cute, I haven't done this style of making providers before. Is there an advantage over a function that produces an assertion?

Reply:
This is the correct method for providing new runners to `tasty`. This also gives me full control over output, as well as the ability to add any `tasty` options I see fit.